File: src/main/scala/ru/ifmo/model/Surface.scala
package ru.ifmo.model

import java.io.File
import scala.io.Source
import ru.ifmo.docking.lip.LipophilicityCalculator
import ru.ifmo.docking.geometry

class Surface(
               pc: Array[Point],
               nc: Array[Vector],
               sc: Array[(Int, Int, Int)],
               ec: Array[Double],
               lc: Array[Double]
               ) {

  val points: Array[Point] = pc
  val normals: Array[Vector] = nc
  val electrostaticPotentials: Array[Double] = ec
  val lipophilicPotentials: Array[Double] = lc
  val faces: Array[(Int, Int, Int)] = sc

  lazy val meshResolution: Double = {
    val edgesLengthSum = faces.foldLeft(0.0) {
      case (acc, (a, b, c)) => acc + (points(a) distance points(b)) + (points(b) distance points(c)) + (points(a) distance points(c))
    }
    edgesLengthSum / ((faces.length + points.length - 2) * 2) //euler formula for edges count
  }

  def transform(matrix: Matrix): Surface = {
    val transformedPoints = points.map(p => GeometryTools.transformPoint(p, matrix))
    val transformedNormals = normals.map(n => GeometryTools.transformVector(n, matrix))
    new Surface(transformedPoints, transformedNormals, faces, electrostaticPotentials, lipophilicPotentials)
  }

}

object Surface {

  def read(surfaceFile: File, electrostaticFile: File, fiPotentials: File, pdbFile: File): Surface = {
    val points: Array[Point] = readPoints(surfaceFile)
    println(s"Read ${points.length} points")
    new Surface(
      points,
      readNormals(surfaceFile),
      readSurfaces(surfaceFile),
      readCsv(electrostaticFile),
      computeLipophilicity(points, fiPotentials, pdbFile)
    )
  }

  private def computeLipophilicity(points: Array[Point], fiPotentials: File, pdbFile: File): Array[Double] = {
    val calculator: LipophilicityCalculator = LipophilicityCalculator.construct(pdbFile, fiPotentials)
    points.map(p => calculator.compute(new geometry.Point(p.x, p.y, p.z)))
  }

  private def readPoints(surfaceFile: File): Array[Point] = {
    val lines: Iterator[String] = Source.fromFile(surfaceFile).getLines().filter(s => s.startsWith("v "))
    val tokens: Iterator[Array[String]] = lines map (_.split("\\s+"))
    val points: Iterator[Point] = tokens map (t => new Point(t(1).toDouble, t(2).toDouble, t(3).toDouble))
    points.toArray
  }

  private def readNormals(surfaceFile: File): Array[Vector] = {
    val lines: Iterator[String] = Source.fromFile(surfaceFile).getLines().filter(s => s.startsWith("vn "))
    val tokens: Iterator[Array[String]] = lines map (_.split("\\s+"))
    val vectors: Iterator[Vector] = tokens map (t => new Vector(t(1).toDouble, t(2).toDouble, t(3).toDouble))
    val normals: Iterator[Vector] = vectors map (_.unite)
    normals.toArray
  }

  private def readSurfaces(surfaceFile: File): Array[(Int, Int, Int)] = {
    val lines: Iterator[String] = Source.fromFile(surfaceFile).getLines().filter(s => s.startsWith("f "))
    val tokens: Iterator[Array[String]] = lines map (_.split("\\s+"))
    val surfaces: Iterator[(Int, Int, Int)] = tokens map (t => (t(1).takeWhile(_.isDigit).toInt, t(2).takeWhile(_.isDigit).toInt, t(3).takeWhile(_.isDigit).toInt))
    surfaces.toArray
  }

  private def readCsv(surfaceFile: File): Array[Double] = {
    val lines: Iterator[String] = Source.fromFile(surfaceFile).getLines()
    val values: Iterator[Double] = lines map (s => s.substring(s.lastIndexOf(',') + 1).toDouble)
    values.toArray
  }
}

File: src/main/java/ru/ifmo/docking/model/Surface.java
package ru.ifmo.docking.model;

import ru.ifmo.docking.calculations.LipophilicityCalculator;
import ru.ifmo.docking.geometry.Geometry;
import ru.ifmo.docking.geometry.Point;
import ru.ifmo.docking.geometry.Vector;
import ru.ifmo.docking.util.IOUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class Surface {
    public final String name;
    public final List<Point> points;
    public final List<Vector> normals;
    public final List<Face> faces;
    public final List<Double> lipophilicity;
    public final List<Double> electricity;

    public Surface(String name,
                   List<Point> points,
                   List<Vector> normals,
                   List<Face> faces,
                   List<Double> lipophilicity,
                   List<Double> electricity) {
        this.name = name;
        this.points = Collections.unmodifiableList(points);
        this.normals = Collections.unmodifiableList(normals);
        this.faces = Collections.unmodifiableList(faces);
        this.lipophilicity = Collections.unmodifiableList(lipophilicity);
        this.electricity = Collections.unmodifiableList(electricity);
    }

    public static Surface read(String name, File pdbFile, File surfaceFile, File electricityFile, File fiPotentials) {
        List<Point> points = new ArrayList<>();
        List<Vector> normals = new ArrayList<>();
        List<Face> faces = new ArrayList<>();

        IOUtils.linesStream(surfaceFile)
                .map(line -> line.split("\\s+"))
                .forEach(tokens -> {
                    switch (tokens[0]) {
                        case "v":
                            points.add(new Point(Double.parseDouble(tokens[1]), Double.parseDouble(tokens[2]), Double.parseDouble(tokens[3])));
                            break;
                        case "vn":
                            normals.add(new Vector(Double.parseDouble(tokens[1]), Double.parseDouble(tokens[2]), Double.parseDouble(tokens[3])).unite());
                            break;
                        case "f":
                            faces.add(new Face(Integer.parseInt(tokens[1]), Integer.parseInt(tokens[2]), Integer.parseInt(tokens[3])));
                            break;
                    }
                });


        LipophilicityCalculator calculator = LipophilicityCalculator.construct(pdbFile, fiPotentials);
        List<Double> lipophilicity = points.stream()
                .map(calculator::compute)
                .collect(Collectors.toList());

        List<Double> electricity = IOUtils.linesStream(electricityFile)
                .map(line -> Double.valueOf(line.substring(line.lastIndexOf(',') + 1)))
                .collect(Collectors.toList());

        return new Surface(name, points, normals, faces, lipophilicity, electricity);
    }

    public double getDiameter() {
        double result = 0.0;
        for (Point p1 : points) {
            for (Point p2 : points) {
                result = Math.max(result, Geometry.distance(p1, p2));
            }
        }
        return result;
    }

    public double getAverageEdgeLength() {
        double sum = 0.0;
        for (Face face : faces) {
            sum += Geometry.distance(points.get(face.p1 - 1), points.get(face.p2 - 1));
            sum += Geometry.distance(points.get(face.p2 - 1), points.get(face.p3 - 1));
            sum += Geometry.distance(points.get(face.p3 - 1), points.get(face.p1 - 1));
        }
        return sum / (faces.size() * 3);
    }


    public static class Face {
        public final int p1;
        public final int p2;
        public final int p3;

        public Face(int p1, int p2, int p3) {
            this.p1 = p1;
            this.p2 = p2;
            this.p3 = p3;
        }
    }
}
